src/utils.ts

```diff
@@ -148,10 +168,76 @@ export async function fetchFlutterServerPort({
          return forwardedPort!;
       } catch (e) {
          if (portReleaseCallback) {
-            await portReleaseCallback(udid, systemPort!);
+            await portReleaseCallback(udid, forwardedPort!);
+         }
+         if (portForwardCallback) {
+            await portForwardCallback(udid, systemPort!, devicePort);
+         }
+         try {
+            log.info(
+               `Checking if flutter server is running on port ${devicePort}`,
+            );
+            await waitForFlutterServer(
+               forwardedPort!,
+               packageName,
+               flutterCaps,
+            );
+            log.info(
+               `Flutter server is successfully running on port ${devicePort}`,
+            );
+            return forwardedPort!;
+         } catch (e) {
+            if (portReleaseCallback) {
+               await portReleaseCallback(udid, systemPort!);
+            }
          }
+         devicePort++;
       }
-      devicePort++;
    }
    return null;
 }
+
+export function isW3cCaps(caps: any) {
+   if (!_.isPlainObject(caps)) {
+      return false;
+   }
+
+   const isFirstMatchValid = () =>
+      _.isArray(caps.firstMatch) &&
+      !_.isEmpty(caps.firstMatch) &&
+      _.every(caps.firstMatch, _.isPlainObject);
+   const isAlwaysMatchValid = () => _.isPlainObject(caps.alwaysMatch);
+   if (_.has(caps, 'firstMatch') && _.has(caps, 'alwaysMatch')) {
+      return isFirstMatchValid() && isAlwaysMatchValid();
+   }
+   if (_.has(caps, 'firstMatch')) {
+      return isFirstMatchValid();
+   }
+   if (_.has(caps, 'alwaysMatch')) {
+      return isAlwaysMatchValid();
+   }
+   return false;
+}
+
+export function attachAppLaunchArguments(parsedCaps: any, ...caps: any) {
+   const capsToUpdate = [...caps].find(isW3cCaps);
+   const platformName: string | undefined = parsedCaps['platformName'];
+   const systemPort: string | undefined = parsedCaps['flutterSystemPort'];
+   console.log(
+      'Caps to update:',
+      systemPort,
+      capsToUpdate,
+      `${capsToUpdate.alwaysMatch['appium:flutterSystemPort']}`,
+   );
+   if (platformName && systemPort && platformName.toLowerCase() == 'ios') {
+      const args = [
+         `--port=${capsToUpdate.alwaysMatch['appium:flutterSystemPort']}`,
```

Comment:
different clients might send the cap in any or both firstMatch and alwaysMatch objects